packages/viewer/src/components/viewer-pin-label/vertex-pin-label.tsx

```diff
@@ -0,0 +1,242 @@
+
+  @Element()
+  private hostEl!: HTMLElement;
+
+  @State()
+  private focused = false;
+
+  private inputEl?: HTMLInputElement;
+
+  public constructor() {
+    if (this.pin?.attributes?.labelText != null) {
+      this.value = this.pin.attributes.labelText;
+    } else {
+      this.value = '';
+    }
+  }
+
+  /**
+   * Gives focus to the the component's internal text input.
+   */
+  @Method()
+  public async setFocus(): Promise<void> {
+    // HTMLInputElement.focus() doesn't exist in tests.
+    if (typeof this.inputEl?.focus === 'function') {
+      this.inputEl?.focus();
+    }
+  }
+
+  @Watch('focused')
+  protected watchFocusChange(): void {
+    this.labelChanged.emit();
+  }
+
+  protected componentDidLoad(): void {
+    const resize = new ResizeObserver(() => {
+      this.labelChanged.emit();
+    });
+
+    if (this.hostEl != null) {
+      resize.observe(this.hostEl);
+    }
+  }
+
+  protected componentDidRender(): void {
+    if (this.focused && this.inputEl != null) {
+      this.setFocus();
+    }
+  }
+
+  protected render(): JSX.Element {
+    const onUpdatePin = (updatedPin: TextPin): void => {
+      this.pinController?.setEntity(updatedPin);
+    };
+    const pointerDownAndMove = (pointerDown: PointerEvent): Disposable => {
+      this.pinController?.setSelectedPinId(this.pin?.id);
+
+      const pointerMove = (event: PointerEvent): void => {
+        if (this.elementBounds != null) {
+          const point = getMouseClientPosition(event, this.elementBounds);
+          const myUpdatedPin: TextPin | undefined = isTextPin(this.pin)
+            ? new TextPin(this.pin.id, this.pin.worldPosition, this.pin.point, {
+                labelPoint: translatePointToRelative(point, this.elementBounds),
+                labelText: this.pin.attributes.labelText,
+              })
+            : undefined;
+
+          if (myUpdatedPin) {
+            onUpdatePin(myUpdatedPin);
+          }
+        }
+      };
+
+      const dispose = (): void => {
+        window.removeEventListener('pointermove', pointerMove);
+        window.removeEventListener('pointerup', pointerUp);
+      };
+
+      const pointerUp = (pointerUp: PointerEvent): void => {
+        const distnaceBetweenStartAndEndPoint = Point.distance(
+          Point.create(pointerDown.clientX, pointerUp.clientY),
+          Point.create(pointerUp.clientX, pointerUp.clientY)
+        );
+
+        if (distnaceBetweenStartAndEndPoint <= 2) {
+          this.focused = true;
+        }
+        dispose();
+      };
+
+      window.addEventListener('pointermove', pointerMove);
+      window.addEventListener('pointerup', pointerUp);
+
+      return {
+        dispose,
+      };
+    };
+
+    const screenPosition =
+      isTextPin(this.pin) &&
+      this.elementBounds != null &&
+      this.pin.attributes.labelPoint != null
+        ? translatePointToScreen(
+            this.pin.attributes.labelPoint,
+            this.elementBounds
+          )
+        : undefined;
```

Comment:
Similarly, this looks like it only recomputes when `this.pin` or `this.elementBounds` changes, so it might be nice to have a `@Watch` on those to recompute this value and set state versus having to recompute every render